hydra-node/test/Hydra/Events/RotationSpec.hs

```diff
@@ -0,0 +1,107 @@
+module Hydra.Events.RotationSpec where
+
+import Hydra.Prelude
+import Test.Hydra.Prelude
+
+import Data.List qualified as List
+import Hydra.Chain.ChainState (ChainSlot (..))
+import Hydra.Events (EventSink (..), HasEventId (..), getEvents)
+import Hydra.Events.Rotation
+import Hydra.Ledger.Simple (SimpleChainState (..), simpleLedger)
+import Hydra.Logging (showLogsOnFailure)
+import Hydra.Node (hydrate)
+import Hydra.NodeSpec (createMockSourceSink, inputsToOpenHead, notConnect, primeWith, runToCompletion)
+import Test.Hydra.Tx.Fixture (testEnvironment)
+import Test.QuickCheck (Positive (..), (==>))
+import Test.QuickCheck.Instances.Natural ()
+
+spec :: Spec
+spec = parallel $ do
+  describe "Node" $ do
+    -- Set up a hydrate function with fixtures curried
+    let setupHydrate action =
+          showLogsOnFailure "NodeSpec" $ \tracer -> do
+            let testHydrate = hydrate tracer testEnvironment simpleLedger SimpleChainState{slot = ChainSlot 0}
+            action testHydrate
+    around setupHydrate $ do
+      it "rotates while running" $ \testHydrate -> do
+        failAfter 1 $ do
+          now <- getCurrentTime
+          let initialChainState = SimpleChainState{slot = ChainSlot 0}
+          let checkpointer = mkChechpointer initialChainState now
+          let logId = 0
+          let rotationConfig = RotateAfter 4
+          eventStore <- createMockSourceSink
+          rotatingEventStore <- mkRotatedEventStore rotationConfig checkpointer logId eventStore
+          testHydrate rotatingEventStore []
+            >>= notConnect
+            >>= primeWith inputsToOpenHead
+            >>= runToCompletion
+          rotatedHistory <- getEvents (fst rotatingEventStore)
+          length rotatedHistory `shouldBe` 2
+      it "consistent state after restarting with rotation" $ \_testHydrate -> do
+        pendingWith "TODO"
+    prop "a rotated an non rotated node have consistent state" $ pendingWith "TODO"
+
+  describe "Rotation algorithm" $ do
+    prop "rotates on startup" $
+      \(Positive x, Positive y) ->
+        (y > x) ==> do
+          eventStore@(eventSource, eventSink) <- createMockSourceSink
+          let totalEvents = toInteger y
+          let events = TrivialEvent <$> [1 .. fromInteger totalEvents]
```

Comment:
Could: just generate two `[TrivialEvent]` directly. Then you also don't need to generate positive numbers (there are no negative number of list items).

Something like `prop ... $ \toRotate additionalEvents -> let totalEvents = toRotate <> additionalEvents; let rotationConfig = RotateAfter (length toRotate) ...`